watchtower/wtclient/queue.go

```diff
@@ -0,0 +1,554 @@
+				// unhandled task is assigned to leftOverItem3
+				// so that it can be handled by the stop method.
+				case <-q.quit:
+					q.leftOverItem3 = &task
+					return
+				}
+			}
+
+			// Send a signal that there is a new item in the main
+			// disk queue.
+			select {
+			case q.newDiskItemSignal <- struct{}{}:
+			case <-q.quit:
+				return
+			default:
+			}
+
+		// If the mode is modeMem, then try feed it to the
+		// manageMemQueueInput handler via the un-buffered inputChan
+		// channel. We wrap it in an internal task so that we can find
+		// out if manageMemQueueInput successfully handled the item. If
+		// it did, we continue in modeMem and if not, then we switch to
+		// modeDisk so that we can persist the item to the disk queue
+		// instead.
+		case modeMem:
+			success := make(chan bool)
+			it := &internalTask[T]{
+				task:    task,
+				success: success,
+			}
+
+			select {
+			// Try feed the task to the manageMemQueueInput handler.
+			// The handler, if it does take the task, is guaranteed
+			// to respond via the success channel of the task to
+			// indicate if the task was successfully added to the
+			// memQueue.
+			case q.inputChan <- it:
+
+			// If the queue is quit at this moment, then the
+			// unhandled task is assigned to leftOverItem3 so that
+			// it can be handled by the stop method.
+			case <-q.quit:
+				q.leftOverItem3 = &task
+
+				return
+
+			default:
+				// Was not accepted. So maybe the mode changed.
+				goto checkMode
+			}
+
+			// If we get here, it means that the manageMemQueueInput
+			// handler took the task. It is guaranteed to respond
+			// via the success channel, so we wait for that response
+			// here.
+			s := <-success
+			if s {
+				continue
+			}
+
+			// If the task was not successfully handled by
+			// manageMemQueueInput, then we switch to modeDisk so
+			// that the task can be persisted in the disk queue
+			// instead.
+			q.setMode(modeDisk)
+
+			goto checkMode
+		}
+	}
+}
+
+// manageMemQueueInput manages which items should be fed onto the buffered
+// memQueue. If the queue is then in modeDisk, then the handler will read new
+// tasks from the disk queue until it is empty. After that, it will switch
+// between reading from the input channel or the disk queue depending on the
+// queue mode.
+func (q *DiskOverflowQueue[T]) manageMemQueueInput() {
+	defer func() {
+		close(q.memQueue)
+		q.wg.Done()
+	}()
+
+	// If the queue is in modeDisk, then the memQueue is fed with tasks
+	// from the disk queue until it is empty.
+	if q.getMode() == modeDisk {
+		for {
+			task, err := q.db.Pop()
+			if errors.Is(err, wtdb.ErrEmptyQueue) {
+				q.setMode(modeMem)
+				break
+			} else if err != nil {
+				log.Errorf("could not load next task from " +
+					"disk. Retrying.")
+
+				select {
+				case <-time.After(dbErrorBackoff):
+					continue
+				case <-q.quit:
+					return
+				}
+			}
+
+			select {
+			case q.memQueue <- task:
```

Comment:
Still trying to wrap my head around the three goroutines and the left over items. One thing we could do is to implement a `Peek()` method in the DB that returns a taks (or err if empty) but doesn't pop it yet. That way we could remove it from the DB here, once we've successfully added it to the memory queue and avoid needing to store it as left over item.